docs/architecture/cloud-native/rest-grpc.md

```diff
@@ -1,69 +1,347 @@
+      {
+      }
+      /// <summary>Creates a new client for Basket that uses a custom <c>CallInvoker</c>.</summary>
+      /// <param name="callInvoker">The callInvoker to use to make remote calls.</param>
+      public BasketClient(grpc::CallInvoker callInvoker) : base(callInvoker)
+      {
+      }
+      /// <summary>Protected parameterless constructor to allow creation of test doubles.</summary>
+      protected BasketClient() : base()
+      {
+      }
+      /// <summary>Protected constructor to allow creation of configured clients.</summary>
+      /// <param name="configuration">The client configuration.</param>
+      protected BasketClient(ClientBaseConfiguration configuration) : base(configuration)
+      {
+      }
+      ...
+    }
+  }
+```
+
+**Figure 4-25** gRPC client plumbing code
+
+Next, you create a custom client class that will invoke the remote gRPC service. Figure 4-26 shows a set of client classes from eShop for invoking gRPC services.
+
+```csharp
+public class BasketService : IBasketService
+{
+    private readonly UrlsConfig _urls;
+    private readonly ILogger<BasketService> _logger;
+
+    public BasketService(HttpClient httpClient, IOptions<UrlsConfig> config, ILogger<BasketService> logger)
+    {
+        _urls = config.Value;
+        _logger = logger;
+    }
+
+    public async Task<BasketData> GetById(string id)
+    {
+        return await GrpcCallerService.CallService(_urls.GrpcBasket, async channel =>
+        {
+            var client = new Basket.BasketClient(channel);
+            _logger.LogDebug("grpc client created, request = {@id}", id);
+            var response = await client.GetBasketByIdAsync(new BasketRequest { Id = id });
+            _logger.LogDebug("grpc response {@response}", response);
+                return MapToBasketData(response);
+            });
+        }
+    }
+    ...
+}
+
+public static class GrpcCallerService
+{
+    public static async Task<TResponse> CallService<TResponse>(string urlGrpc, Func<GrpcChannel, Task<TResponse>> func)
+    {
+        var channel = GrpcChannel.ForAddress(urlGrpc);
+
+        Log.Information("Creating grpc client base address urlGrpc ={@urlGrpc}, BaseAddress={@BaseAddress} ", urlGrpc, channel.Target);
+
+        try
+        {
+            return await func(channel);
+        }
+        catch (RpcException e)
+        {
+            Log.Error("Error calling via grpc: {Status} - {Message}", e.Status, e.Message);
+            return default;
+        }
+    }
+    ...
+}
+```
+
+**Figure 4-26** Custom gRPC client code
+
+Note in the previous figure how *GetById* method in the *BasketService* class creates an anonymous method. The method is passed into the generic *CallService* method in the *GrpcCallerService*. Importantly, you can see how *CallService* consumes the system-generated *Basket.BasketClient* and *gRPChannel* gRPC plumbing code to invoke remote calls.
+
+### gRPC server implementation
```

Comment:
As mentioned earlier, Server & Client implementation details can be avoided in this book.

Reply:
Removed and refactored.

Reply:
Removed